### ISSUE TITLE
FreeSync not working with managed windows

### ISSUE THREAD

**Issue by mserajnik** [herbstluftwm/herbstluftwm#1347]
I'm experiencing the issue that FreeSync isn't working under herbstluftwm (using no compositor) using a Radeon RX 6900 XT GPU, while it works just fine under KDE Plasma. See my [initial issue](https://gitlab.freedesktop.org/drm/amd/-/issues/1659) in the AMD DRM project that I made before finding out it's herbstluftwm-specific for more details.

After figuring that out, I did some more testing and noticed that it does work with unmanaged windows, e.g.:

```bash
hc rule class='love' manage=off
```

With this rule, the [VRRTest](https://github.com/Nixola/VRRTest) window is unmanaged, causing FreeSync to work.
However, that is not really ideal since now the window apparently doesn't get focus (and therefore doesn't get keyboard inputs) as well as it also not being possible to access the rest of the desktop anymore (I'm effectively stuck in that one fullscreen window); I would still need to be able to switch workspaces and use the rest of the desktop as usual while running a game in fullscreen with FreeSync.

Given that KDE Plasma is able to have managed fullscreen windows with FreeSync working (at least that's what I assume since they integrate seamlessly), I am not sure if this is a bug or limitation in herbstluftwm, or if I am simply not aware how to configure this properly.

On a sidenote: when I was still using a GeForce GTX 1080 Ti about half a year ago, G-SYNC (or rather NVIDIA's implementation of FreeSync, since it's just a "G-SYNC compatible" FreeSync monitor) was working just fine under herbstluftwm, but maybe that was due to a implementation difference compared to AMD's FreeSync.

**Comment by t-wissmann** [comment]
Hi, thanks for the report and sorry for the trouble! This will be hard to reproduce because I don't have any radeon GPUs here. I don't think that the fullscreen state is a real issue here, because the fullscreen isn't anything special (it is essentially maximized, without decoration, and hiding panels).

The main difference between normal windows and unmanaged windows in this regard is that unmanaged windows are not reparented. Window reparenting means that the actual client window is swallowed by another window which shows the window decorations.

My first guess is that the issue here is the [EWMH `_NET_WM_SYNC_REQUEST` protocol](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm46841634376672) which is not implemented by hlwm.

**Comment by mserajnik** [comment]
@t-wissmann 

Thank you for the quick response!

> The main difference between normal windows and unmanaged windows in this regard is that unmanaged windows are not reparented. Window reparenting means that the actual client window is swallowed by another window which shows the window decorations.

I think this might probably be the reason, as a kernel developer also wrote:

> Sounds like herbstluftwm somehow prevents page flipping from being used for fullscreen windows. E.g. maybe it enables backing store for all windows, or reparents windows to a parent window with a different visual, which results in automatic compositing.

But I am not well-versed enough here to tell how hard that would be to change.

I understand that this is a very niche use case, considering herbstluftwm itself is probably already quite niche. So I could absolutely understand if you say this is out of scope for the project.

**Comment by t-wissmann** [comment]
If hlwm does not adhere to some protocol it should be fixed :-) (so far I didn't see any necessity to implement the Sync-Request thing, but this changed with the present issue).

> ... or reparents windows to a parent window with a different visual, which results in automatic compositing.

This is interesting. Only since commit cecfccdb33494c59c69cb74b4b17b32d0b8630af (#1186), hlwm always uses a 32-bit visual to allow true transparency in window decorations. But actually I would have guessed that plasma does something similar. But just to exlcude that your issue wasn't triggered there: can you check whether the issue already existed in commits f25d52fae7a128d226fbe00a6c4b2445f99b525f and also tag/commit `v0.9.2`? (there was another change regarding visuals between those two commits: 696afd4f1e4bc7860e4478820f6c69bb06220c76 (#1178) and I'd like to double check that this commit didn't cause the issue)

**Comment by mserajnik** [comment]
I wasn't able to build f25d52fae7a128d226fbe00a6c4b2445f99b525f and 79fcd0d1ab8e8a221418ee340331b16b5710b0bc (the `v0.9.2` release commit) with GCC 11 (uploaded the build log [here](https://gist.github.com/mserajnik/8a232ed7109a9f677e8f8ba48a06c619), just in case you are interested), so I temporarily downgraded to GCC 10 to build them, which worked (GCC `10.2.0`, to be precise; also downgraded CMake to `3.19.7`, so it's compatible).

I can confirm that FreeSync is also not working with either of those two commits.

**Comment by mserajnik** [comment]
Also, in case that is useful information: `_VARIABLE_REFRESH(CARDINAL) = 1` seems to get set (I assume by the [accompanying Xorg driver for GPUs using the amdgpu kernel driver](https://gitlab.freedesktop.org/xorg/driver/xf86-video-amdgpu)) for windows that should/"want to" support FreeSync.

I can see this property is getting set regardless of the window being managed or unmanaged by herbstluftwm.

**Comment by mserajnik** [comment]
@t-wissmann 

Sorry, I have to correct myself:

FreeSync does indeed work just fine and as I would expect it to in both f25d52fae7a128d226fbe00a6c4b2445f99b525f and 79fcd0d1ab8e8a221418ee340331b16b5710b0bc (the `v0.9.2` release commit).

This then also explains why I had no trouble with my NVIDIA card months ago; because it only broke in a newer commit.

A friend of mine (who has more experience with the Xorg protocol(s) than me) did some tinkering to get a temporary workaround going for me (with unparenting fullscreen windows automatically) and that's when I noticed that it's not enough to simply run `herbstclient reload` after rebuilding the executable, as that won't re-run herbstluftwm, but just reload the configuration instead.

Sorry for the confusion and the trouble; in practice that means for me, I can just use f25d52fae7a128d226fbe00a6c4b2445f99b525f as a workaround in the meantime, so please don't feel pressured from my side to fix this regression. And I am, of course, willing to test potential fixes (or help with debugging in general). :)

**Comment by t-wissmann** [comment]
Thanks for the clarification! I will look for a way to fix this issue.

**Comment by t-wissmann** [comment]
So it seems the best way to fix this is to not reparent clients in fullscreen state, as done in [matchbox](http://git.yoctoproject.org/cgit/cgit.cgi/libmatchboxwm2/plain/matchbox/mb-wm-window-type-simple.c) and [awesome](https://gitmemory.com/issue/awesomeWM/awesome/2764/491312944).

**Comment by t-wissmann** [comment]
In the current git version, there is now a `decorated` attribute (but there is no rule for it yet). If you add a keybinding, you can toggle client decorations at run time:
```
hc keybind $Mod-Shift-d attr clients.focus.decorated toggle
```
Is the sync issue fixed for you if you disable the decorations?

**Comment by EuklidAlexandria** [comment]
Unfortunately, disabling the decorations does not fix the sync issue for me. It still uses blitting instead of flipping.

**Comment by mserajnik** [comment]
> Is the sync issue fixed for you if you disable the decorations?

@t-wissmann 

Sorry for not replying to your question. I had some real life stuff coming up around that time and completely forgot about this. After that period, I switched to macOS for work reasons.

I'm now about to switch back to Linux in a few weeks (I have ordered a new notebook) and while thinking about what desktop environment I might want to use, herbstluftwm sprung to my mind and I finally got reminded of this issue again.

If I do end up using herbstluftwm on the notebook, I'll give the `decorated` attribute a try and report back; but it seems like it might not work based on @EuklidAlexandria's feedback from 2023? I assume there haven't been any other developments on that front since then?

**Comment by t-wissmann** [comment]
No, there hasn't been much development in hlwm in general, because unfortately, I was also very busy with real life stuff and work.
